### PR TITLE
JavascriptPlayerProxy: expose formatted key in user defined notation type

### PR DESCRIPTION
**Pull request by ne0phyte**
Hi,

first PR here and I have basically no knowledge of C++ and QT, but I believe this change is pretty straightforward.
The change exposes the track bpm and key string values to controller scripts via the `JavascriptPlayerProxy`.

I needed this as a prerequisite for triggering a search for related tracks from within controller scripts that I am working on, but this might be useful on its own.

### CODE REVIEW

File: src/controllers/scripting/javascriptplayerproxy.cpp [hunk 1]
```diff
@@ -84,6 +92,8 @@ void JavascriptPlayerProxy::slotTrackLoaded(TrackPointer pTrack) {
     emit yearChanged(m_pCurrentTrack->getYear());
     emit trackNumberChanged(m_pCurrentTrack->getTrackNumber());
     emit trackTotalChanged(m_pCurrentTrack->getTrackTotal());
+    emit bpmChanged(m_pCurrentTrack->getBpmText());
```
Comment (JoergAtGithub) [inline src/controllers/scripting/javascriptplayerproxy.cpp]: For numerical values, we use control objects, not strings. 
And therefore we've already already a control object for this purpose: https://manual.mixxx.org/2.6/en/chapters/appendix/mixxx_controls#control-[ChannelN]-bpm

File: src/controllers/scripting/javascriptplayerproxy.cpp [hunk 2]
```diff
@@ -84,6 +92,8 @@ void JavascriptPlayerProxy::slotTrackLoaded(TrackPointer pTrack) {
...
+    emit keyChanged(m_pCurrentTrack->getKeyText());
```
Comment (JoergAtGithub) [inline src/controllers/scripting/javascriptplayerproxy.cpp]: Please add a test case for this, that tests with all supporte Key Notations

**Comment by JoergAtGithub** [comment]
Welcome at Mixxx!
As a first-time contributor we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

**Review by JoergAtGithub — CHANGES_REQUESTED** [review]

File: src/controllers/scripting/javascriptplayerproxy.cpp (hunk 1 above)
Comment (ne0phyte) [reply src/controllers/scripting/javascriptplayerproxy.cpp]: Oh. I don't know how I missed that. Sorry for your troubles.
Since there already is a way to get both `file_bpm` and `file_key`, is there even a point in adding the string representation to the player proxy or should I just close this PR?

Comment (JoergAtGithub) [reply src/controllers/scripting/javascriptplayerproxy.cpp]: For BPM I see no need for a string representation.
But for controllers with display, it would make sense to get a string of the key, in the key notation, that the user selected.
But the later has many options and partly uses special characters. This requires a proper testcase, as we need to ensure backward compatibility of this API in future. Do you want to work on this?

Comment (ne0phyte) [reply src/controllers/scripting/javascriptplayerproxy.cpp]: I kind of did already :)

I removed the bpm property and wrote a test case that verifies that the correct key notations based on the user configured key notation type and/or custom notation reaches the JS context.

After my test failed I realized that `Track::getKeyText()` always uses the `KeyNotation::Custom` so I rewrote the `JavascriptPlayerProxy` to use the `[Library]key_notation` CO state to format the key with `KeyUtils` explicitly in the selected notation.

I was not sure whether it'd be better to split it up into smaller tests (per Notation?) or create separate `TestSpec` lists to get rid of the constantly repeating `KeyNotation`. But since tests are rarely beautiful I left it as it is for now.

It was a bit of trial & error and a lot of looking at other tests so I'm not sure if this is what you had in mind.

**EDIT**: I'm not sure why the tests are failing in the pipeline. They ran successfully locally last night. I'm looking into it.

**Comment by ne0phyte** [comment]
> Welcome at Mixxx! As a first-time contributor we need you to sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and comment here when you have done so. It gives us permission to distribute your contribution under the GPL v2 or later license and the Apple Mac App Store. It is also helpful for us to have contact information for contributors in case we may need it in the future.

Added. Since I saw the question in other PRs: if this ends up being merged you can use my real name in the contributors list.

File: src/controllers/scripting/javascriptplayerproxy.cpp (hunk 1 above)
Comment (JoergAtGithub) [reply src/controllers/scripting/javascriptplayerproxy.cpp]: How could it ran throuh locally? It seems to compare the name instead of the value.
Otherwise the aproach looks good.

Comment (ne0phyte) [reply src/controllers/scripting/javascriptplayerproxy.cpp]: I figured out what happened.
Initially (and incorrectly) I used `[Master]key_notation` in both the test and the `JavascriptPlayerProxy`. Since the `JavascriptPlayerProxy` uses a `ControlProxy` it received the CO that was created in the test case and my tests worked. 

The last thing I did last night was swap that to use `[Library]` after I realized that the proxy output didn't actually respond to changing the Mixxx configuration. I only tested the change in Mixxx and didn't re-run the test. 

I updated the test to use a `ControlProxy` instead of creating a new CO, since the Library (which creates the CO in its constructor) is initialized in `SetUp`. I also updated both places to use the const `mixxx::library::prefs::kKeyNotationConfigKey`.

This is probably all obvious to you, I'm just explaining what I did since it wasn't obvious to me :)

I am not sure what you mean by comparing the name instead of the value? I set the key notation CO, set the `Keys` to the loaded `Track` and then check the output on the JS side against the expected string. 

The test just completed successfully on my machine and I confirmed by logging the key from a controller script and changing the Mixxx settings that all key notations print as expected on the JS side.

File: src/test/controllerscriptenginelegacy_test.cpp [hunk 3]
```diff
@@ -996,6 +999,212 @@ TEST_F(ControllerScriptEngineLegacyTest, JavascriptPlayerProxy) {
...
+            {ChromaticKey::G_SHARP_MINOR, KeyNotation::ID3v2, "G#m"},
+            {ChromaticKey::A_MINOR, KeyNotation::ID3v2, "Am"},
+            {ChromaticKey::B_FLAT_MINOR, KeyNotation::ID3v2, "Bbm"},
+            {ChromaticKey::B_MINOR, KeyNotation::ID3v2, "Bm"},
+    };
+
+    TrackPointer pTrack = m_pPlayerManager->getDeck(0)->getLoadedTrack();
+    ASSERT_NE(pTrack, nullptr);
+
+    for (const auto& tc : cases) {
```
Comment (JoergAtGithub) [inline src/test/controllerscriptenginelegacy_test.cpp]: We don't use auto, were the type is not obvious. Please use the explicit type name here.
The variable name tc is not self-descriptive

**Comment by JoergAtGithub** [comment]
Please add the new method also to the TypeScript declaration of JavascriptPlayerProxy in res/controllers/engine-api.d.ts

File: src/test/controllerscriptenginelegacy_test.cpp (hunk 3 above)
Comment (ne0phyte) [reply src/test/controllerscriptenginelegacy_test.cpp]: Fixed. Added the explicit type and renamed `cases` -> `testCases` and `tc` -> `testCase` for clarity.

**Comment by ne0phyte** [comment]
> Please add the new method also to the TypeScript declaration of JavascriptPlayerProxy in res/controllers/engine-api.d.ts

Done. I also added another connection in the `JavascriptPlayerProxy` to call `keyChanged` when the user changes the key notation in the settings to reflect the change on a controller immediately.

I added another test case `JavascriptPlayerProxy_KeyNotation_keyChanged` for that behavior and tested manually by logging to console from a controller script and changing the settings.

**Question:**
There is one last edge case that I haven't figured out yet. When the notation is set to Custom and the Custom notation is edited and saved, it doesn't trigger a change of the `key_notation` CO and therefore won't update the proxy.
My gut feeling says that it is not worth handling that edge case explicitly as users probably don't change the Custom notation regularly and the controller script will get the update sooner or later anyways.

However, if you see that as a hard requirement I'll look into it. Let me know what you think.

**Comment by JoergAtGithub** [comment]
We only consider such exceptions, if there is a real issue with implementing it always working.
Do you know, where the change event signal get lost?

**Comment by ne0phyte** [comment]
I think I figured it out. In `Library` I changed the key_notation CO initialization to set `bIgnoreNops` to false. 

The problem was that I didn't find a dedicated signal for changes to the custom notation. The `DlgPrefKey` sets the selected `KeyNotation` and custom notation strings on apply/ok, but setting `KeyNotation::Custom` again didn't result in a call in the proxy.

Now the JS `player.keyChanged` listeners get called when the custom notation is edited in the settings. 

EDIT: I force pushed because I had a typo in the commit message of the fix. I didn't change any code or previous commits.

**Comment by JoergAtGithub** [comment]
Great, could you please add a minimal testcase to test this change of the custom notation.

**Comment by ne0phyte** [comment]
I added another test case for that.

Since this is still relying on the key_notation CO changing (which is already covered) and I didn't see any UI tests I assumed that you wanted me to just call the two methods the same way `DlgPrefKey` does, updating the custom notation `QMap` and checking that the JS object receives the updated value.

**Review by JoergAtGithub — APPROVED** [review]
LGTM! Thank you!
